src/java/org/apache/cassandra/index/sai/utils/OrderIterator.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.Closeable;
+
+/**
+ * An iterator over {@link ScoredPrimaryKey} ordered by score in either ascending or descending order.
+ */
+public abstract class OrderIterator extends AbstractIterator<ScoredPrimaryKey> implements Closeable
```

Comment:
Maybe it is just me, but the name `OrderIterator` doesn't tell me much. 
Maybe `ScoredPrimaryKeyIterator` would be a better name? Just thinking loud. I guess the idea behind `OrderIterator` was that it comes from iterating the index that provides the `ORDER BY` order?

Reply:
I am not confident on the naming here. I went with `OrderIterator` because it is for the ORDER BY clause, and we already have the `RangeIterator` everywhere. I am open to other names. `ScoredPrimaryKeyIterator` could work.

Reply:
I agree with Piotr, OrderIterator sounds like it's going to give me a templated interface for generic ordering somehow

On a related note: do we even need a new interface here or can we just use CloseableIterator<SPK> and Collections.emptyIterator()?

Reply:
I switched `OrderIterator` to `ScoredPrimaryKeyIterator`, and I replaced the `ScoredRowIdIterator` with `CloseableIterator<ScoredRowId>`. I might switch `ScoredPrimaryKeyIterator` to `CloseableIterator<ScoredPrimaryKey>`. A dedicated interface does give us some minor niceties:

1. We can document the intention of the type.
2. `AbstractIterator<ScoredPrimaryKey>` feels mildly easier to implement because you only need to implement `computeNext()` instead of `next()` and `hasNext()`.
3. `ScoredPrimaryKeyIterator` is shorter than `CloseableIterator<ScoredPrimaryKey>`

I'll continue to think about it, and might push up a change.